application/features/bootstrap/UnitTestsContext.php

```diff
@@ -0,0 +1,170 @@
+    }
+
+    protected function createMfa($type, $verified = 1, $user = null)
+    {
+        $user = $user ?? $this->tempUser;
+
+        $mfa = new Mfa();
+        $mfa->user_id = $user->id;
+        $mfa->type = $type;
+        $mfa->verified = $verified;
+
+        Assert::true($mfa->save(), "Could not create new mfa.");
+        $user->refresh();
+        $this->mfaId = $mfa['id'];
+        $this->mfa = $mfa;
+    }
+
+    protected function createMethod($value, $verified = 1, $user = null)
+    {
+        $user = $user ?? $this->tempUser;
+
+        $method = new Method();
+        $method->user_id = $user->id;
+        $method->value = $value;
+        $method->verified = $verified;
+
+        Assert::true($method->save(), "Could not create new method.");
+        $user->refresh();
+    }
+
+    /**
+     * @Given A user with :verMeth verified method(s), :unverMeth unverified method(s), :verMfa verified mfa(s), and :unverMfa unverified mfa(s)
+     */
+    public function aUserWithVerifiedMethodsVerifiedMfas($verMeth, $unverMeth, $verMfa, $unverMfa)
+    {
+        $this->tempUser = $this->createNewUserInDatabase('method_tester');
+
+        for ($i = 0; $i < $verMeth; $i++) {
+            $this->createMethod('verified' . $i . '@example.org', 1);
+        }
+
+        for ($i = 0; $i < $unverMeth; $i++) {
+            $this->createMethod('unverified' . $i . '@example.org', 0);
+        }
+
+        for ($i = 0; $i < $verMfa; $i++) {
+            $this->createMfa('totp', 1);
+        }
+
+        for ($i = 0; $i < $unverMfa; $i++) {
+            $this->createMfa('totp', 0);
+        }
+    }
+
+    /**
+     * @When I request a list of verified methods
+     */
+    public function iRequestAListOfVerifiedMethods()
+    {
+        $this->methodList = $this->tempUser->getVerifiedMethodOptions();
+    }
+
+    /**
+     * @Then I see a list containing :num method
+     */
+    public function iSeeAListContainingMethod($num)
+    {
+        Assert::count($this->methodList, $num);
+    }
+
+
+    /**
+     * @Given the nag dates are in the past
+     */
+    public function theNagDatesAreInThePast()
+    {
+        $this->tempUser->nag_for_method_after = MySqlDateTime::formatDate(time() - (60 * 60 * 24));
+        $this->tempUser->nag_for_mfa_after = MySqlDateTime::formatDate(time() - (60 * 60 * 24));
+    }
+
+    /**
+     * @When I request the nag state
+     */
+    public function iRequestTheNagState()
+    {
+        $this->nagState = $this->tempUser->getNagState();
+    }
+
+    /**
+     * @Then I see that the nag state is :state
+     */
+    public function iSeeThatTheNagStateIs($state)
+    {
+        Assert::eq($this->nagState, $state);
+    }
+
+    /**
+     * @Then I update the nag dates
+     */
+    public function iUpdateTheNagDates()
+    {
+        $this->tempUser->updateNagDates();
+
+        /*
+         * Force an update the next time getNagState() is called
+         */
+        $this->tempUser->nagState = null;
```

Comment:
Is this something that it would be good to do every time the nag dates are updated (and thus, that it would potentially be worth moving into the `updateNagDates()` method)?

Reply:
Perhaps. It's a question of how it might be used in the future. As of now, it wouldn't make a difference. Since `authenticatedUser` is cloned prior to calling `updateNagDates`, it would be safe to reset the `nagState` in `updateNagdates`. Logic seems to be in favor of doing as you suggest.

Reply:
Done.